src/platforms/linux/daemon/dbusservice.cpp

```diff
@@ -250,54 +251,70 @@ void DBusService::appTerminated(const QString& cgroup, const QString& appId) {
 QString DBusService::runningApps() {
   QJsonArray result;
 
-  for (auto i = m_appTracker->begin(); i != m_appTracker->end(); i++) {
-    const AppData* data = *i;
+  for (const QString& cgroup : m_appTracker->getRunningApps()) {
     QJsonObject appObject;
-    QJsonArray pidList;
-    appObject.insert("appId", QJsonValue(data->appId));
-    appObject.insert("cgroup", QJsonValue(data->cgroup));
-    appObject.insert("rootpid", QJsonValue(data->rootpid));
+    appObject.insert("id", QJsonValue(m_appTracker->getDesktopId(cgroup)));
+    appObject.insert("cgroup", QJsonValue(cgroup));
 
-    for (int pid : data->pids()) {
-      pidList.append(QJsonValue(pid));
-    }
+    AppState state = m_excludedCgroups.value(cgroup, Active);
+    QMetaEnum meta = QMetaEnum::fromType<AppState>();
+    appObject.insert("state", QJsonValue(meta.valueToKey(state)));
 
-    appObject.insert("pids", pidList);
     result.append(appObject);
   }
 
   return QJsonDocument(result).toJson(QJsonDocument::Compact);
 }
 
+void DBusService::setAppState(const QString& desktopId, AppState state) {
+  logger.debug() << "Setting" << desktopId << "to firewall state" << state;
+
+  // When the App is "Active" there is no special manipulation to do.
+  if (state == Active) {
+    m_excludedApps.remove(desktopId);
+    for (const QString& cgroup : m_appTracker->findByDesktopId(desktopId)) {
+      m_wgutils->resetCgroup(cgroup);
+    }
+    return;
+  }
+
+  // Otherwise, apply special handling to any matching control groups.
+  m_excludedApps[desktopId] = state;
+  for (const QString& cgroup : m_appTracker->findByDesktopId(desktopId)) {
+    if (m_excludedCgroups.contains(cgroup)) {
+      m_wgutils->resetCgroup(cgroup);
+    }
+    m_excludedCgroups[cgroup] = state;
+    if (state == Excluded) {
+      // Excluded control groups are given special netfilter rules to direct
+      // their traffic outside of the VPN tunnel.
+      m_wgutils->excludeCgroup(cgroup);
+    }
```

Comment:
Should we check if the cgroup state is already in the expected state before doing this work?

Reply:
That felt like an unnecessary edge case to handle. If the cgroup is already in the expected state then this will simply remove the firewall rules for it, and then re-add them.